File: gke_customization/gke_hrms/report/absence_report/absence_report.js
// Copyright (c) 2024, Gurapps/gke_customization/gke_customization/gke_hrms/report/absence_report/absence_report.jsukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["Absence Report"] = {
    filters: [
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.get_today(),
        },
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company"),
            on_change: function (report) {
                let company = frappe.query_report.get_filter_value("company");
                let branch_filter = frappe.query_report.get_filter("branch");
                let department_filter = frappe.query_report.get_filter("department");

                if (company === "KG GK Jewellers Private Limited") {
                    branch_filter.df.read_only = 1;
                    branch_filter.set_value("");
                    branch_filter.refresh();
                } else {
                    branch_filter.df.read_only = 0;
                    branch_filter.refresh();
                }

                department_filter.set_value("");
                department_filter.refresh();
            },
        },
        {
            fieldname: "branch",
            label: __("Branch"),
            fieldtype: "Link",
            options: "Branch",
            depends_on: "eval:doc.company && doc.company != 'KG GK Jewellers Private Limited'",
            get_query: function () {
                let company = frappe.query_report.get_filter_value("company");
                return company ? { filters: { company } } : {};
            },
        },
        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Department",
            get_query: function () {
                let company = frappe.query_report.get_filter_value("company");
                return company ? { filters: { company: company } } : {};
            },
        },
    ],

    onload: function (report) {
        report.refresh();
    },
};

File: gke_customization/gke_hrms/report/absence_report/absence_report.py
# Copyright (c) 2026, Gurukrupa Export
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    if not filters:
        filters = {}

    if not filters.get("from_date") or not filters.get("to_date"):
        frappe.throw(_("From Date and To Date are required"))

    columns = get_columns()
    data = get_data(filters)
    return columns, data


# -------------------------------------------------------
# Columns
# -------------------------------------------------------

def get_columns():
    return [
        {
            "label": _("Date"),
            "fieldname": "attendance_date",
            "fieldtype": "Date",
            "width": 130,
        },
        {
            "label": _("Employee"),
            "fieldname": "employee",
            "fieldtype": "Link",
            "options": "Employee",
            "width": 280,
        },
        {
            "label": _("Company"),
            "fieldname": "company",
            "fieldtype": "Link",
            "options": "Company",
            "width": 160,
        },
        {
            "label": _("Department"),
            "fieldname": "department",
            "fieldtype": "Link",
            "options": "Department",
            "width": 160,
        },
        {
            "label": _("Designation"),
            "fieldname": "designation",
            "fieldtype": "Data",
            "width": 140,
        },
        {
            "label": _("Shift"),
            "fieldname": "shift",
            "fieldtype": "Link",
            "options": "Shift Type",
            "width": 150,
        },
        {
            "label": _("Last Check-in"),
            "fieldname": "last_checkin",
            "fieldtype": "Datetime",
            "width": 210,
        },
        {
            "label": _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 90,
        },
    ]


def get_data(filters):
    conditions = get_conditions(filters)

    query = f"""
        SELECT 
            a.attendance_date,
            e.employee,
            e.employee_name,
            e.department,
            e.company,
            e.designation,
            a.shift,

            (
                SELECT MAX(ec.time)
                FROM `tabEmployee Checkin` ec
                WHERE ec.employee = e.employee
                AND ec.docstatus < 2
            ) AS last_checkin,

            CASE 
            WHEN la.name IS NOT NULL THEN la.leave_type
            ELSE 'Absent'
            END AS status


        FROM `tabAttendance` a
        INNER JOIN `tabEmployee` e ON e.employee = a.employee
        LEFT JOIN `tabLeave Application` la
            ON la.employee = a.employee
            AND la.status = 'Approved'
            AND a.attendance_date BETWEEN la.from_date AND la.to_date

        WHERE
            a.attendance_date BETWEEN %(from_date)s
                AND LEAST(%(to_date)s, DATE_SUB(CURDATE(), INTERVAL 1 DAY))
            AND DAYOFWEEK(a.attendance_date) != 1
            AND a.status IN ('Absent', 'On Leave')
            AND a.docstatus = 1
            AND e.status = 'Active'
            {conditions}

        UNION ALL

        SELECT 
            CURDATE() AS attendance_date,
            e.employee,
            e.employee_name,
            e.department,
            e.company,
            e.designation,
            NULL AS shift,

            (
                SELECT MAX(ec.time)
                FROM `tabEmployee Checkin` ec
                WHERE ec.employee = e.employee
                AND ec.docstatus < 2
            ) AS last_checkin,

            CASE 
                WHEN la.name IS NOT NULL THEN 'LWP'
                ELSE 'Absent'
            END AS status

        FROM `tabEmployee` e
        LEFT JOIN `tabLeave Application` la
            ON la.employee = e.employee
            AND la.status = 'Approved'
            AND CURDATE() BETWEEN la.from_date AND la.to_date

        WHERE
            e.status = 'Active'
            AND DAYOFWEEK(CURDATE()) != 1
            AND CURDATE() BETWEEN %(from_date)s AND %(to_date)s
            AND NOT EXISTS (
                SELECT 1
                FROM `tabAttendance` a
                WHERE a.employee = e.employee
                AND a.attendance_date = CURDATE()
                AND a.docstatus = 1
            )
            AND NOT EXISTS (
                SELECT 1
                FROM `tabEmployee Checkin` ec
                WHERE ec.employee = e.employee
                AND ec.docstatus < 2
            AND DATE(ec.time) = CURDATE()
            )
            {conditions}

        ORDER BY
            attendance_date DESC,
            department ASC
    """

    return frappe.db.sql(query, filters, as_dict=True)


# -------------------------------------------------------
# Conditions
# -------------------------------------------------------

def get_conditions(filters):
    conditions = []

    if filters.get("company"):
        conditions.append("e.company = %(company)s")

    if filters.get("branch"):
        conditions.append("e.branch = %(branch)s")

    if filters.get("department"):
        conditions.append("e.department = %(department)s")

    if conditions:
        return " AND " + " AND ".join(conditions)

    return ""
